Port to Python3  (alpha, active issues, GCI)

srevinsaju opened the pull request:
# Port to Python3 - Wikipedia Activit

## Known Issues
* [ ] `EOL`'d dependencies:
`mwlib` is assumed to be obtained from https://github.com/gwillem/magento-malware-scanner
This is no longer maintained from 2018. 
* [ ] Can't complete test until _mwscan is fixed, for that, the mwlib should be either removed or ported, or updated from src.
@quozl

quozl commented:
mwlib is from https://github.com/godiard/mwlib

quozl commented:
mwlib is likely upstream at https://github.com/pediapress/mwlib

quozl commented:
acb8ff1 adds files at top level of activity, which would be misleading to new developers.

srevinsaju commented:
@quozl, I will remove off the unwanted files, but Its not working though :(
It needs a developer with C language knowledge

srevinsaju commented:
@quozl , currently mwlib porting is not possible as what pediapress tells, the only option is to:
(Its not only the String error, the devlopers mention of other errors too, it requires manual compiling of cython to continue porting to Python3 (`mwlib`))
(i) Continue using python only for mwlib and port the activity to python3
(iI) Remove mwlib to https://github.com/earwig/mwparserfromhell

quozl commented:
Alright then, you've explained why our older mwlib has to be ported.  Ensure that explanation is in the commit message of the port of our older mwlib.

srevinsaju commented:
@quozl the creators of mwlib has commented that its currently not possible to port mwlib to Python3. For now they are using https://github.com/earwig/mwparserfromhell .
I will have to discard this pull request, and will have to start from scratch, the imports are too many, and its need is not well commented causing me to go :sob:

quozl commented:
Where did they comment?  Why not possible?  Everything is possible with code, eventually.

srevinsaju commented:
@quozl it seems they are working on python3.
a message by @ckepper was on the python3 port

https://github.com/pediapress/mwlib/issues/66#issuecomment-568417893

should I try to port it, or wait for the official release

quozl commented:
If it were me, I'd offer to help.  If you don't feel you can do that in the time available to you, then set it aside.

srevinsaju commented:
So far, the commit https://github.com/godiard/wikipedia-activity/pull/1/commits/e7b0e8e7673050fe194a0d8b8779d6023ea632e7 is the last functional version. The `mwlib-simple` is not a good alternative.

quozl commented:
How is this pull request going?

538fde2 seems to add redundant parentheses, as if 2to3 is run twice.

srevinsaju commented:
I have probably ran 2to3 twice on some files, buy I will fix it. good news is mwlib is fixed by removing the unneeded functions. please review the .re files because I don't know if what I have done is right

srevinsaju commented:
@quozl please review this: A raw HACK https://github.com/godiard/wikipedia-activity/pull/1/commits/5be2eea39c8d4134547ec324ba27ccb87d63ad85

quozl commented:
You'd have to explain 5be2eea, as I don't understand it.

Please confirm you have tested creating a new activity bundle from a Wikipedia download, and that the bundle does work on Sugar with Python3?

srevinsaju commented:
@quozl , When I try to compile mwlib after partially porting it, It asks for a constructor or a destrutor
```no constructor or destructor or type conversion found before init_mwscan```
This caused me to refer different articles, but I had no clue in trying to understand this stuff. So I just commented out the `init_mwscan` part and the error no longer exists while building. 
But still, it gave some more errors like `_mwscan` cannot import error, (for obvious reasons), so I removed the import too, and put it in a try except. The activity depends on the sugar-web activity, and needs porting there. WIP. I wanted to know if what I did on https://github.com/godiard/wikipedia-activity/commit/5be2eea39c8d4134547ec324ba27ccb87d63ad85 is right. 

PS:
I have changed PyUnicodeObject to PyASCII object because of unicode -> string conversions, and this is valid. Idk about the rest

quozl commented:
Thanks.  I guess the port has to be complete rather than partial.  Python is written in C, so it shouldn't be too difficult to figure it all out.  I've not the time to do it myself though, yet.  If you aren't able to proceed, and nobody else can help, then we'll hang onto this pull request for a while.

quozl commented:
I suggest reading or searching the C source code for Python.

srevinsaju commented:
yes @quozl I will make the port complete in a while, but changes to a C code is hard to review for me. I will make the activity working in a while. 
I have removed some parts of the reason files which had the word `extern`. I hope nothing bad happens if I remove it

srevinsaju commented:
> I suggest reading or searching the C source code for Python.

Yes, I read the src and changed `PyUnicode` to `PyASCII`. and that fixed the error. i am only suspicious of removing extern in the reason

quozl commented:
Okay.  _extern_ in C means that the variable is to be allocated in memory by another source file.  If something that was _extern_ cannot be found, it means it was removed from some other source file.  If that source file is part of Python, then it means you have to find if it was replaced by something else, or some API.  So you'd download Python 2 source code and Python 3 source code and see what has changed in the area of the variable definition.

Then, since Python is maintained in git, you could go find the commit message that matches the change in order to understand what has to be changed in the code you are responsible for.

srevinsaju commented:
@quozl thanks for that!

quozl commented:
Working?  Great!  If there's any changes to the documentation here or on the Wiki explaining how to rebuild the activity with a new extract from Wikipedia, please let us all know.

srevinsaju commented:
@quozl, Yes, its working very rudimentarily. Some more fixes needed, need to remove mwlib, add `py-wikimarkup`. Currently, the activity looks for the file and goes to en.wikipedia ... /<topic> directly, as mwlib cannot extract the files, `py-wikimarkup` looks promising. :)

The wikipedia-activity was Python2, GTK2, and used sugar instead of sugar3 . The server commands had changed overtime.

srevinsaju commented:
@quozl I have updated the checklist for your reference on work left to complete. Thanks

srevinsaju commented:
@walterbender @Hrishi1999 @chimosky Please review. Thanks

quozl commented:
Thanks.  Reviewed, not tested.

* [ ] seems to be a lot of whitespace change to the .re files, such that the real changes are invisible,
* [ ] the port of mwlib is incomplete; there's commented code, and mwlib is still referenced by server; please remove any code not needed,
* [ ] some other files, like pylru.py, have been rewrapped, again can't see the real changes if any; or did it come from somewhere else?,
* [ ] can you include pijnu as source rather than a tar.gz?

Overall I'm concerned that the removal of mwlib makes the creation of a new activity bundle a very expensive CPU operation.  If it can be ported instead, it should be better.  On the other hand, we've failed to keep the activity bundle up to date anyway, and a reasonable update rate may be once or twice every year.

srevinsaju commented:
@quozl I can remove all the mwlib resources. Should I proceed with that then.

Regarding CPU usage: to get the best performance at the lowest CPU usage, it is important to port the `mwlib` and not use the `mwparser`. The new `mwparser` has a lot of problems, example tables and images. However, there is no development in the `pediapress/mwlib`, and I do not have any idea of the C files too. Is there anything I can do for this PR?

quozl commented:
Sure.  Keep mwlib.  Learn C.  It's not hard.  Python is based on C.  GIve yourself about ten hours to get started, and about a week to build the rest of the knowledge required.  It is common in software engineering to learn a language only enough to fix something.  In this situation, although I know C, and I know Python, what I don't know yet is how to write an extension for Python in C, nor how to do it for both Python 2 and Python 3.  That's the skill set needed for mwlib porting.  It is all empirical though; nothing needs to be a mystery.  For myself I don't need to do this **yet**, because I've added Python 2 support to OLPC OS next release.

quozl commented:
Chapter 8 of [Supporting Python 3](http://python3porting.com/pdfs/SupportingPython3-screen-1.0-latest.pdf) covers _Migrating C Extensions_.

srevinsaju commented:
Yes, I read that during GCI, and I tried them too, but maybe, because I didn't do them good. I had learnt C momentarily, like the syntax, but understanding pointers were way beyond my understanding. I guess, unless I have some real life program to work on C with Python libs so that I can understand them more properly. I may try in future, maybe in a few months,
